### PR TITLE
PUBDEV-3827: fix PCA scoring history and importance.  Added Runit tes…

### PR DESCRIPTION
**Pull request by wendycwong**
…t derived from Angela JIRA.

PUBDEV-3797: Fix PCA scoring history and importance.  Fixed Python interface to add PCA method Randomized.

PUBDEV-3827: fix PCA scoring history and importance.  Add compute_metrics to legal list of parameters for PCA in Python.

PUBDEV-3827: fix PCA scoring history and importance.  Added pyunit test to check scoring history and importance for every PCA mode.

PUBDEV-3827: Fix PCA scoring history and importance.  Fixed bug in java.  Parameters total_variance was not calculated.  Improved pyunit test to check out all transform mode and all PCA_methods.

PUBDEV-3827. Fix PCA scoring history and importance.  Correct calculation of total_variance.

PUBDEV-3827: fix pca scoring history and importance.  Fix total_variance calculation when PCA method is glrm.

PUBDEV-3827: fix PCA scoring history importance.  Fixed tolerance level.  Generate more printing for debugging.

PUBDEV-3827: fix PCA scoring history and importance.  Added Runit test and cleaning up.

PUBDFEV-3827: PCA scoring history and importance.  Clean up code to calculate importance in one spot for PCA/GLRM.  Fixed bugs in calculating importance for all PCA modes.

PUBDEV-3827: add scoring history and importance to PCA.  Fixed Runit test and now it is just like the Pyunit test except that we can compare R PCA result with H2O PCA results when transform is STANDARDIZE.

### CODE REVIEW

**Comment by wendycwong** [comment]
@h2o-ops please test!

File: h2o-core/src/main/java/water/util/TwoDimTable.java [hunk 1]
```diff
@@ -219,6 +219,16 @@ else if ((long)val==val)
   public int getColDim() { return colHeaders.length; }
 
   /**
+   * Need to change table header when we are calling GLRM from PCA.
+   *
+   * @param newHeader: String containing new table header.
+   */
+  public void setTableHeader(String newHeader) {
+    if ((newHeader != null) && (newHeader.length() > 0))
+    this.tableHeader = newHeader;
```
Comment (st-pasha) [inline h2o-core/src/main/java/water/util/TwoDimTable.java]: bad indentation
also: `StringUtils.isNullOrEmpty()`

File: h2o-py/h2o/transforms/decomposition.py [hunk 2]
```diff
@@ -10,9 +10,10 @@ class H2OPCA(H2OEstimator):
     def __init__(self, model_id=None, k=None, max_iterations=None, seed=None,
                  transform=("NONE", "DEMEAN", "DESCALE", "STANDARDIZE", "NORMALIZE"),
                  use_all_factor_levels=False,
-                 pca_method=("GramSVD", "Power", "GLRM"),
+                 pca_method=("GramSVD", "Power", "GLRM", "Randomized"),
```
Comment (st-pasha) [inline h2o-py/h2o/transforms/decomposition.py]: pca_method="GramSVD"

File: h2o-py/h2o/transforms/decomposition.py [hunk 3]
```diff
@@ -41,8 +42,12 @@ def __init__(self, model_id=None, k=None, max_iterations=None, seed=None,
...
               the SVD using local matrix algebra.
+        :param bool ignore_const_cols: If true, will ignore constant columns.  Default is True.
+        :param bool impute_missing:  whether to impute NA/missing values.
+        :param bool compute_metrics: whether to compute metrics on training data.  Default to True
 
         :returns: A new instance of H2OPCA.
+
         """
         super(H2OPCA, self).__init__()
         self._parms = locals()
```
Comment (st-pasha) [inline h2o-py/h2o/transforms/decomposition.py]:
```
from h2o.utils.typechecks import Enum
assert_is_type(pca_method, Enum("GramSVD", "Power", "GLRM", "Randomized"))
```

Comment (st-pasha) [inline h2o-py/h2o/transforms/decomposition.py]: same for `transform`

Comment (st-pasha) [inline h2o-py/h2o/transforms/decomposition.py]: remove the 
```
self._parms["pca_method"] = "GramSVD" if isinstance(pca_method, tuple) else pca_method
self._parms["transform"] = "NONE" if isinstance(transform, tuple) else transform
```

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 4]
```diff
@@ -1356,6 +1356,70 @@ def show_test_results(test_name, curr_test_val, new_test_val):
...
+
+    assert size1==size2, "The two H2OTwoDimTables are of different size!"
+    assert num_comparison<=size1, "H2OTwoDimTable do not have all the attributes specified in col_header_list."
+
+    for ind in range(num_comparison):
+        col_name = col_header_list[ind]
+        next_name=False
+
+        for name_ind1 in range(len(table1.cell_values)):
+            if not(col_name==str(table1.cell_values[name_ind1][0])):
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: !=

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 5]
```diff
@@ -1356,6 +1356,70 @@ def show_test_results(test_name, curr_test_val, new_test_val):
...
+    This method compares two H2OTwoDimTables and verify that their difference is less than value set in tolerance. It
+    is probably an overkill for I have assumed that the order of col_header_list may not be in the same order as
+    the values in the table.cell_values[ind][0].  In addition, I do not assume an order for the names in the
+    table.cell_values[ind][0] either for there is no reason for an order to exist.
+
+    :param table1: H2OTwoDimTable to be compared
+    :param table2: the other H2OTwoDimTable to be compared
+    :param col_header_list: list of strings denote names that we can the comparison to be performed
+    :param tolerance: default to 1e-6
+    :param check_sign: bool, determine if the sign of values are important or not.  For eigenvectors, they are not.
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: Better make sign conversion utility as a separate method; and ensure to flip sign on the entire vector at once, not on individual components.
Eg: (1, 2, 3) == (-1, -2, -3) but not (1, -2, 3)

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 6]
```diff
@@ -1356,6 +1356,70 @@ def show_test_results(test_name, curr_test_val, new_test_val):
...
+
+                        if (type(val1) == float) and (type(val2) == float):
+                            if check_sign:
+                                compare_val = abs(table1.cell_values[name_ind1][indC]-
+                                                  table2.cell_values[name_ind2][indC])
+                            else:
+                                compare_val = abs(abs(table1.cell_values[name_ind1][indC])-
+                                                  abs(table2.cell_values[name_ind2][indC]))
+                            compare_val_ratio = compare_val/max(abs(val1), abs(val2))
+                            if (compare_val > tolerance) and (compare_val_ratio > tolerance):
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: same as
max(compare_val, compare_val_ratio) > tolerance

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 7]
```diff
@@ -1356,6 +1356,70 @@ def show_test_results(test_name, curr_test_val, new_test_val):
...
+    is probably an overkill for I have assumed that the order of col_header_list may not be in the same order as
+    the values in the table.cell_values[ind][0].  In addition, I do not assume an order for the names in the
+    table.cell_values[ind][0] either for there is no reason for an order to exist.
+
+    :param table1: H2OTwoDimTable to be compared
+    :param table2: the other H2OTwoDimTable to be compared
+    :param col_header_list: list of strings denote names that we can the comparison to be performed
+    :param tolerance: default to 1e-6
+    :param check_sign: bool, determine if the sign of values are important or not.  For eigenvectors, they are not.
+    :return: True if comparison succeed and raise an error if comparison failed for whatever reason
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: return nothing, but throw an exception if comparison fails

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 8]
```diff
@@ -1356,6 +1356,70 @@ def show_test_results(test_name, curr_test_val, new_test_val):
         print(pass_string)
         return 0
 
+def equal_H2OTwoDimTable(table1, table2, col_header_list, tolerance=1e-6, check_sign=False):
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: rename to `assert_TwoDimTable_equal`

File: h2o-py/tests/testdir_algos/pca/pyunit_PUBDEV_3827_scoringHistory_importance.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                                                                                      "empty."
+  assert len(glrmPCA._model_json["output"]["scoring_history"].cell_values) > 0, "PCA Scoring history setting " \
+                                                                                "pca_method to GLRM is empty."
+
+
+
+if __name__ == "__main__":
+  pyunit_utils.standalone_test(pca_scoring_history_importance)
+else:
+  pca_scoring_history_importance()
```
Comment (st-pasha) [inline h2o-py/tests/testdir_algos/pca/pyunit_PUBDEV_3827_scoringHistory_importance.py]: in python indentation is 4 spaces

File: h2o-r/tests/testdir_algos/pca/runit_pca_pubdev_3694_Rotterdam_large.R [hunk 10]
```diff
@@ -22,7 +22,8 @@ check.pca.widedata.gramSVD <- function() {
     h2o_pca <- h2o.prcomp(df, k = ranks, x = x, transform = 'STANDARDIZE', seed=12345)      # train H2O PCA
 
     # the eigenvectors and eigvalues calculated from R and H2O are close but not equal
-    isFlipped1 <- checkPCAModel(h2o_pca, pcaR, tolerance=1)
+    browser()
```
Comment (st-pasha) [inline h2o-r/tests/testdir_algos/pca/runit_pca_pubdev_3694_Rotterdam_large.R]: remove

File: h2o-core/src/main/java/water/util/ArrayUtils.java [hunk 11]
```diff
@@ -1637,8 +1639,97 @@ public static boolean equalsAny(long value, long...lhs) {
...
+    List<String> colFormat = new ArrayList<>();
+    Boolean seeField = (averSEE != null);
+    Boolean eigField = (eigenVectorIndex != null);
+
+    for (String fieldName:fieldNames) {
+      switch(fieldName) {
+        case "Timestamp": colHeaders.add("Timestamp"); colTypes.add("string"); colFormat.add("%s"); break;
+        case "Duration":  colHeaders.add("Duration"); colTypes.add("string"); colFormat.add("%s"); break;
+        case "Iteration": colHeaders.add("Iteration"); colTypes.add("long"); colFormat.add("%d"); break;
+        default: colHeaders.add(fieldName); colTypes.add("double"); colFormat.add("%d");
```
Comment (wendycwong) [inline h2o-core/src/main/java/water/util/ArrayUtils.java]: Done

File: h2o-core/src/main/java/water/util/ArrayUtils.java [hunk 12]
```diff
@@ -1637,8 +1639,97 @@ public static boolean equalsAny(long value, long...lhs) {
...
+            colFormat.toArray(new String[0]),
+            "");
+
+    for (int row = 0; row < rows; row++) {
+      int col = 0;
+      assert (row < table.getRowDim());
+      assert (col < table.getColDim());
+
+      // take care of Timestamp, Duration, Iteration.
+      DateTimeFormatter fmt = DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss");
```
Comment (wendycwong) [inline h2o-core/src/main/java/water/util/ArrayUtils.java]: Done.  Use a HashMap.

File: h2o-core/src/main/java/water/util/TwoDimTable.java (hunk 1 above)
Comment (wendycwong) [inline h2o-core/src/main/java/water/util/TwoDimTable.java]: Done

File: h2o-py/h2o/transforms/decomposition.py (hunk 2 above)
Comment (wendycwong) [inline h2o-py/h2o/transforms/decomposition.py]: Done

File: h2o-py/h2o/transforms/decomposition.py (hunk 3 above)
Comment (wendycwong) [inline h2o-py/h2o/transforms/decomposition.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 8 above)
Comment (wendycwong) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 7 above)
Comment (wendycwong) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 4 above)
Comment (wendycwong) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 6 above)
Comment (wendycwong) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: It should be min(compare_val, compare_val_ration) > tolerance.

Done

File: h2o-py/tests/testdir_algos/pca/pyunit_PUBDEV_3827_scoringHistory_importance.py (hunk 9 above)
Comment (wendycwong) [inline h2o-py/tests/testdir_algos/pca/pyunit_PUBDEV_3827_scoringHistory_importance.py]: The indentation is 4 spaces.

File: h2o-r/tests/testdir_algos/pca/runit_pca_pubdev_3694_Rotterdam_large.R (hunk 10 above)
Comment (wendycwong) [inline h2o-r/tests/testdir_algos/pca/runit_pca_pubdev_3694_Rotterdam_large.R]: Done

**Comment by vpatryshev** [comment]
(just curious) Do any pr tests pass at all? Or do people merge with tests failing?

File: h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+     * 2. All other fields are double.
+     *
+     * The following field will be generated for you automatically: Duration and Iteration.
+     *
+     * @param scoreTable: HashMap containing column headers and arraylist containing the history of values collected.
+     * @param tableName: title/name of your scoring table
+     * @param startTime:  time your model building job was first started.
+     * @return: TwoDimTable containing the scoring history.
+     */
+    public static TwoDimTable createScoringHistoryTableDR(HashMap<String, ArrayList> scoreTable, String tableName,
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java]: LinkedHashMap is an ordered version of a HashMap

File: h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        ArrayList<String> otherTableEntries = new ArrayList<String>();
+
+        if (scoreTable.keySet().contains("Timestamp")) {
+            colHeaders.add("Timestamp"); colTypes.add("string"); colFormat.add("%s");
+            colHeaders.add("Duration"); colTypes.add("string"); colFormat.add("%s");
+            colHeaders.add("Iteration"); colTypes.add("long"); colFormat.add("%d");
+        }
+
+        for (String fieldName:scoreTable.keySet()) {
+            switch(fieldName) {
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java]: if (!fieldName.equals("Timestamp")) {
   ....
}

File: h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+                new String[rows],
+                colHeaders.toArray(new String[0]),
+                colTypes.toArray(new String[0]),
+                colFormat.toArray(new String[0]),
+                "");
+
+        for (int row = 0; row < rows; row++) {
+            int col = 0;
+            assert (row < table.getRowDim());
+            assert (col < table.getColDim());
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java]: col = 0 here always
first assert can be moved outside

File: h2o-py/h2o/transforms/decomposition.py [hunk 16]
```diff
@@ -41,15 +44,29 @@ def __init__(self, model_id=None, k=None, max_iterations=None, seed=None,
...
         self._parms = {k: v for k, v in self._parms.items() if k != "self"}
-        self._parms["pca_method"] = "GramSVD" if isinstance(pca_method, tuple) else pca_method
-        self._parms["transform"] = "NONE" if isinstance(transform, tuple) else transform
-
+        try:
+            assert_is_type(pca_method, Enum("GramSVD", "Power", "GLRM", "Randomized"))
+            self._parms["pca_method"]=pca_method
+        except:
```
Comment (st-pasha) [inline h2o-py/h2o/transforms/decomposition.py]: remove try-except
```
assert_is_type(pca_method, Enum("GramSVD", "Power", "GLRM", "Randomized"))
self._parms["pca_method"] = pca_method
```

File: h2o-py/h2o/transforms/decomposition.py [hunk 17]
```diff
@@ -41,15 +44,29 @@ def __init__(self, model_id=None, k=None, max_iterations=None, seed=None,
...
+            self._parms="GramSVD"
+
+        try:
+            assert_is_type(transform, Enum("NONE", "DEMEAN", "DESCALE", "STANDARDIZE", "NORMALIZE"))
+            self._parms["transform"]=transform
+        except:
+            self._parms["transform"]="NONE"
```
Comment (st-pasha) [inline h2o-py/h2o/transforms/decomposition.py]: same here

File: h2o-py/h2o/transforms/decomposition.py [hunk 18]
```diff
@@ -109,8 +123,17 @@ def __init__(self, nv=None, max_iterations=None, transform=None, seed=None,
...
+        try:
+            assert_is_type(svd_method, Enum("GramSVD", "Power", "GLRM", "Randomized"))
+            self._parms["svd_method"] = svd_method
+        except:
+            self._parms["svd_method"] = "GramSVD"
+        try:
+            assert_is_type(transform, Enum("NONE", "DEMEAN", "DESCALE", "STANDARDIZE", "NORMALIZE"))
+            self._parms["transform"]=transform
+        except:
+            self._parms["transform"]="NONE"
```
Comment (st-pasha) [inline h2o-py/h2o/transforms/decomposition.py]: and here

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 19]
```diff
@@ -1357,6 +1357,73 @@ def show_test_results(test_name, curr_test_val, new_test_val):
...
+
+                # now we have the col header names, do the actual comparison
+                if str(table1.cell_values[name_ind1][0])==str(table2.cell_values[name_ind2][0]):
+                    for indC in range(1, len(table2.cell_values[name_ind2])):
+                        val1 = table1.cell_values[name_ind1][indC]
+                        val2 = table2.cell_values[name_ind2][indC]*flip_sign_vec[indC]
+
+                        if (type(val1) == float) and (type(val2) == float):
+                            compare_val = abs(val1-val2)
+                            compare_val_ratio = compare_val/max(abs(val1), abs(val2))
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]:
```
compare_val_ratio = abs(val1 - val2) / max(1, abs(val1), abs(val2))
```
then compare_val not needed

File: h2o-r/tests/testdir_algos/pca/runit_pca_pubdev_3694_Rotterdam_large.R (hunk 10 above)
Comment (st-pasha) [inline h2o-r/tests/testdir_algos/pca/runit_pca_pubdev_3694_Rotterdam_large.R]: nope

File: h2o-py/h2o/transforms/decomposition.py (hunk 3 above)
Comment (wendycwong) [inline h2o-py/h2o/transforms/decomposition.py]: Done

File: h2o-py/tests/testdir_algos/pca/pyunit_PUBDEV_3827_scoringHistory_importance.py (hunk 9 above)
Comment (wendycwong) [inline h2o-py/tests/testdir_algos/pca/pyunit_PUBDEV_3827_scoringHistory_importance.py]: Fixed

File: h2o-r/tests/testdir_algos/pca/runit_pca_pubdev_3694_Rotterdam_large.R (hunk 10 above)
Comment (wendycwong) [inline h2o-r/tests/testdir_algos/pca/runit_pca_pubdev_3694_Rotterdam_large.R]: Done

File: h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java (hunk 13 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java]: Done

File: h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java (hunk 14 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java]: Done.

File: h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java (hunk 15 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/util/DimensionReductionUtils.java]: Done

File: h2o-py/h2o/transforms/decomposition.py (hunk 16 above)
Comment (wendycwong) [inline h2o-py/h2o/transforms/decomposition.py]: Done

File: h2o-py/h2o/transforms/decomposition.py (hunk 17 above)
Comment (wendycwong) [inline h2o-py/h2o/transforms/decomposition.py]: Done

File: h2o-py/h2o/transforms/decomposition.py (hunk 18 above)
Comment (wendycwong) [inline h2o-py/h2o/transforms/decomposition.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 19 above)
Comment (wendycwong) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: This is nice.  Done.  Thanks.

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by wendycwong** [comment]
@h2o-ops please test!

**Comment by wendycwong** [comment]
wendy_jenkin runs are all green.

**Pull request merged**